src/gwc/src/main/java/org/geoserver/gwc/layer/GeoServerTileLayer.java

```diff
@@ -579,21 +584,30 @@ protected final void sendTileRequestedEvent(ConveyorTile tile) {
         }
     }
 
-    protected ConveyorTile getMetatilingReponse(
-            ConveyorTile tile, final boolean tryCache, final int metaX, final int metaY)
+    protected ConveyorTile getMetatilingResponse(
+            ConveyorTile conveyorTile, final boolean tryCache, final int metaX, final int metaY)
             throws GeoWebCacheException, IOException {
 
-        if (tryCache && tryCacheFetch(tile)) {
-            return finalizeTile(tile);
+        if (tryCache && tryCacheFetch(conveyorTile)) {
+            return finalizeTile(conveyorTile);
         }
 
-        final GeoServerMetaTile metaTile = createMetaTile(tile, metaX, metaY);
-        Lock lock = null;
+        final GeoServerMetaTile metaTile = createMetaTile(conveyorTile, metaX, metaY);
+
+        /* ****************** Acquire lock on metatile ******************* */
+        final Lock metaTileLock =
+                GWC.get().getLockProvider().getLock(buildMetaTileLockKey(conveyorTile, metaTile));
+
+        /* ****************** Acquire lock on individual tile ******************* */
+        final Lock tileLock =
+                GWC.get()
+                        .getLockProvider()
+                        .getLock(buildTileLockKey(conveyorTile, conveyorTile.getTileIndex()));
+
         try {
-            /* ****************** Acquire lock ******************* */
-            lock = GWC.get().getLockProvider().getLock(buildLockKey(tile, metaTile));
-            // got the lock on the meta tile, try again
-            if (tryCache && tryCacheFetch(tile)) {
+
+            // After getting the lock on the meta tile and invidual tile, try again
+            if (tryCache && tryCacheFetch(conveyorTile)) {
```

Comment:
My currently implementation attempts to ensure that the GetMap request for the metatile is only dispatched a single time. My implementation is actually _very close_ to your last proposal. 

You mentioned the following, which is not quite correct:

> In particular, a request thread for another tile might find the tile it wants still hasn't landed on disk, and decides to dispatch the getmap again

The concurrent user request thread will block on the tile lock (obtained by the ongoing async thread) before checking the cache, and when the async thread is complete and unlocks the tile lock, the request thread will find the tile on disk. To help outline that, here is the perspective of the primary request that is able to dispatch the getmap:

1. Check cache before any locks
2. Obtain the meta-tile lock. 
3. After obtaining meta-tile lock, obtain the tile lock
4. Check cache again, after obtaining _both_ the meta-tile and sub-tile lock.
5. If not found, dispatch the getmap
6. Compute the tile inline, execute the others async
7. Async threads obtains locks on all the other tiles
8. Primary thread waits for all tile locks to be obtained before releasing meta-tile lock (using a countdown latch).

The primary request was able to obtain the meta-tile lock _and_ each individual tile lock before releasing the meta-tile lock. The async threads will not release their tile locks until they've saved the tiles to disk. The concurrent user request will eventually obtain the meta-tile lock but then block on their respective tile locks while the async thread is still going. Here is the perspective of a concurrent request that is for a different tile of the same meta-tile:

1. Check cache before any locks.
2. **Blocks** on meta-tile lock initially, but is eventually obtained.
3. **Blocks** on tile lock because an async thread is still working on it, but is eventually obtained.
4. Check cache again, this time it's found because the async thread put it there. No need to dispatch getmap.

This is very similar to your very last proposal with I think only one key difference, from what I can tell. I intentionally don't release the meta-tile lock until I have certainty that all the individual tiles have been locked by the async threads. If there's a concurrent user request, we want it to wait for the async thread and not the other way around because the async threads already have the image whereas the concurrent user request does not (and would need to dispatch a getmap). Therefore, the async thread must be able to obtain the tile lock and the concurrent user request (after obtaining the meta-tile lock) will then block and wait for the async thread.

Pseudo code that could help:

```

getTile(tile){
   
   if (isCached(tile)){
      return getFromCache(tile)
   }

   lock(metatile) // concurrent request will block here if an ongoing getmap is dispatched
   lock(tile) // concurrent request will block here if an async thread is running

   if (isCached(tile)){
      return getFromCache(tile)
   } else {
     dispatchGetMap(metatile);
  
     response = saveTile(tile)
  
     for(tile : metatile.getTiles()){
        async(() -> {
           lock(tile)
           saveTile(tile)
           release(tile)
         })
      }
   
     waitForAllTileLocksToBeObtained()
   
     release(lock)
     release(metatile)
 
     return response  
}
```

I also attached a diagram to help me sort out my own thoughts on the matter. 

![Geoserver Data Flow - WMS PNG Tile with GWC Integration - Metatile Locking (1)](https://github.com/geoserver/geoserver/assets/607425/0e43e101-775e-4e65-86c5-728922675d92)


Please continue to be critical in case I'm overlooking something!

Reply:
As a reviewer, being critical is my job :rofl: 
The current code looks good, and the diagram is indeed useful.